src/blk/kernel/KernelDevice.cc

```diff
@@ -560,22 +560,21 @@ void KernelDevice::_discard_update_threads()
     for(uint64_t i = oldcount; i > newcount; i--)
     {
       discard_threads[i-1]->stop = true;
-      discard_threads[i-1]->detach();
     }
-    discard_threads.resize(newcount);
-
     discard_cond.notify_all();
+    l.unlock();
+    for (uint64_t i = oldcount; i > newcount; i--)
+    {
+      discard_threads[i-1]->join();
```

Comment:
I think one shouldn't access discard_threads out of locked mutex.
So you rather has to make a local copy of threads to stop and resize discard_threads under that lock. Then do the join of ouf the lock.

Reply:
done